src/cargo/core/compiler/mod.rs

```diff
@@ -1162,6 +1168,69 @@ fn features_args(unit: &Unit) -> Vec<OsString> {
     args
 }
 
+/// Generates the `--remap-path-scope` and `--remap-path-prefix` for [RFC 3127].
+/// See also unstable feature [`-Ztrim-paths`].
+///
+/// [RFC 3127]: https://rust-lang.github.io/rfcs/3127-trim-paths.html
+/// [`-Ztrim-paths`]: https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#profile-trim-paths-option
+fn trim_paths_args(
+    cmd: &mut ProcessBuilder,
+    cx: &Context<'_, '_>,
+    unit: &Unit,
+    trim_paths: &TomlTrimPaths,
+) -> CargoResult<()> {
+    if trim_paths.is_none() {
+        return Ok(());
+    }
+
+    // feature gate was checked during mainfest/config parsing.
+    cmd.arg("-Zunstable-options");
+    cmd.arg(format!("-Zremap-path-scope={trim_paths}"));
+
+    let sysroot_remap = {
+        let sysroot = &cx.bcx.target_data.info(unit.kind).sysroot_target_libdir;
+        let commit_hash = &cx.bcx.rustc().commit_hash;
+        let mut remap = OsString::from("--remap-path-prefix=");
+        // TODO(trim-paths): what is the correct prefix for sysroot? does rustup affect this?
+        remap.push(sysroot);
+        remap.push("=");
+        remap.push("/rustc/");
+        remap.push(commit_hash);
+        remap
+    };
+    cmd.arg(sysroot_remap);
+
+    let package_remap = {
+        let pkg_root = unit.pkg.root();
+        let ws_root = cx.bcx.ws.root();
+        let is_local = unit.pkg.package_id().source_id().is_path();
+        let mut remap = OsString::from("--remap-path-prefix=");
+        // Remapped to path relative to workspace root:
+        //
+        // * path dependencies under workspace root directory
```

Comment:
For me, the main question of whether to be relative to workspace or package root is if there'd be ambiguity.  Since we have name+version, I'm guessing not.